pkg/csplugin/broker.go

```diff
@@ -0,0 +1,339 @@
+	"github.com/crowdsecurity/crowdsec/pkg/models"
+	"github.com/crowdsecurity/crowdsec/pkg/protobufs"
+	plugin "github.com/hashicorp/go-plugin"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	PluginProtocolVersion uint   = 1
+	CrowdsecPluginKey     string = "CROWDSEC_PLUGIN_KEY"
+)
+
+type PluginBroker struct {
+	PluginChannel                   chan ProfileAlert
+	alertsByPluginName              map[string][]*models.Alert
+	profileConfigs                  []*csconfig.ProfileCfg
+	pluginConfigByName              map[string]PluginConfig
+	pluginMap                       map[string]plugin.Plugin
+	notificationConfigsByPluginType map[string][][]byte // "slack" -> []{config1, config2}
+	notificationPluginByName        map[string]Notifier
+	watcher                         PluginWatcher
+	pluginKillMethods               []func()
+}
+
+// holder to determine where to dispatch config and how to format messages
+type PluginConfig struct {
+	Type           string        `yaml:"type"`
+	Name           string        `yaml:"name"`
+	GroupWait      time.Duration `yaml:"group_wait"`
+	GroupThreshold int           `yaml:"group_threshold"`
+	MaxRetry       int           `yaml:"max_retry"`
+	TimeOut        time.Duration `yaml:"timeout"`
+
+	Format string `yaml:"format"` // specific to notification plugins
+
+}
+
+type ProfileAlert struct {
+	ProfileID uint
+	Alert     *models.Alert
+}
+
+func (pb *PluginBroker) Init(profileConfigs []*csconfig.ProfileCfg, configPaths *csconfig.ConfigurationPaths) error {
+	pb.PluginChannel = make(chan ProfileAlert)
+	pb.notificationConfigsByPluginType = make(map[string][][]byte)
+	pb.notificationPluginByName = make(map[string]Notifier)
+	pb.pluginMap = make(map[string]plugin.Plugin)
+	pb.pluginConfigByName = make(map[string]PluginConfig)
+	pb.alertsByPluginName = make(map[string][]*models.Alert)
+	pb.profileConfigs = profileConfigs
+	if err := pb.loadConfig(configPaths.NotificationDir); err != nil {
+		return err
+	}
+	err := pb.loadPlugins(configPaths.PluginDir)
+	pb.watcher = PluginWatcher{}
+	pb.watcher.Init(pb.pluginConfigByName, pb.alertsByPluginName)
+	return err
+}
+
+func (pb *PluginBroker) Kill() {
+	for _, kill := range pb.pluginKillMethods {
+		kill()
+	}
+}
+
+func (pb *PluginBroker) Run() {
+	go pb.watcher.Start()
+	for {
+		select {
+		case profileAlert := <-pb.PluginChannel:
+			pb.addProfileAlert(profileAlert)
+
+		case pluginName := <-pb.watcher.C:
+			// this can be ran in goroutine, but then locks will be needed
+			if err := pb.pushNotificationsToPlugin(pluginName); err != nil {
+				log.WithField("plugin:", pluginName).Error(err)
+			}
+			pb.alertsByPluginName[pluginName] = make([]*models.Alert, 0)
+		}
+	}
+}
+
+func (pb *PluginBroker) addProfileAlert(profileAlert ProfileAlert) {
+	for _, pluginName := range pb.profileConfigs[profileAlert.ProfileID].Notifications {
+		if _, ok := pb.pluginConfigByName[pluginName]; !ok {
+			log.Errorf("binary for plugin %s  not found in", pluginName)
```

Comment:
extra space to be removed + `in` not needed

Reply:
I don't follow on this, could you elaborate ?

Reply:
Fixed via https://github.com/crowdsecurity/crowdsec/pull/857/commits/e2ea9833a99a96bd9e7152e870f1353e12d18986